### ISSUE TITLE
threadstats is not threadsafe

### ISSUE THREAD

**Issue by edkorthof** [DataDog/datadogpy#365]
This code appears to be unsafe wrt threads:

```
131     def add_point(self, metric, tags, timestamp, value, metric_class, sample_rate=1, host=None):
132         # The sample rate is currently ignored for in process stuff
133         interval = timestamp - timestamp % self._roll_up_interval
134         key = (metric, host, tuple(sorted(tags)) if tags else None)
135         if key not in self._metrics[interval]:
136             self._metrics[interval][key] = metric_class(metric, tags, host)
137         self._metrics[interval][key].add_point(value)
138
139     def flush(self, timestamp):
140         """ Flush all metrics up to the given timestamp. """
141         interval = timestamp - timestamp % self._roll_up_interval
142         past_intervals = [i for i in self._metrics.keys() if i < interval]
143         metrics = []
144         for i in past_intervals:
145             for m in list(self._metrics.pop(i).values()):
146                 metrics += m.flush(i)
147         return metrics
```

The problem is that flush can run at the same time as add_point, and the operations in add_point are not atomic. If we imagine:

Thread1, ts 19.99999: line 135 checks self._metrics[10] and passes; but this thread is paused
Thread2, ts 20.00000: line 145 pops self._metrics[10]
Thread2, ts 20.00001: line 137 throws KeyError on self._metrics[10]

We saw this error in the wild, ending with:

```
...
  File "/usr/local/lib/python2.7/dist-packages/datadog/threadstats/base.py", line 162, in gauge
    sample_rate=sample_rate, host=host)
  File "/usr/local/lib/python2.7/dist-packages/datadog/threadstats/metrics.py", line 137, in add_point
    self._metrics[interval][key].add_point(value)
KeyError: ('ws_load_check', 'prod-ws193.ec2', ('metric_type:capacity',))
```

I put together a script which can (on a loaded system) throw a similar error. My example script's output:
```
[] ~: ./example.py
Exception in thread Thread-322:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "./example.py", line 57, in add
    aggregator.add_point(metric, tags, ts, value, Gauge)
  File "./example.py", line 37, in add_point
    self._metrics[interval][key].add_point(value)
KeyError: ('a320', None, ('item320',))

Exception in thread Thread-1264:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "./example.py", line 57, in add
    aggregator.add_point(metric, tags, ts, value, Gauge)
  File "./example.py", line 37, in add_point
    self._metrics[interval][key].add_point(value)
KeyError: ('a1262', None, ('item1262',))

Exception in thread Thread-1934:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "./example.py", line 57, in add
    aggregator.add_point(metric, tags, ts, value, Gauge)
  File "./example.py", line 37, in add_point
    self._metrics[interval][key].add_point(value)
KeyError: ('a1932', None, ('item1932',))

Exception in thread Thread-171:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "./example.py", line 57, in add
    aggregator.add_point(metric, tags, ts, value, Gauge)
  File "./example.py", line 37, in add_point
    self._metrics[interval][key].add_point(value)
KeyError: ('a169', None, ('item169',))

Exception in thread Thread-1593:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "./example.py", line 57, in add
    aggregator.add_point(metric, tags, ts, value, Gauge)
  File "./example.py", line 37, in add_point
    self._metrics[interval][key].add_point(value)
KeyError: ('a1591', None, ('item1591',))

Exception in thread Thread-288:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "./example.py", line 57, in add
    aggregator.add_point(metric, tags, ts, value, Gauge)
  File "./example.py", line 37, in add_point
    self._metrics[interval][key].add_point(value)
KeyError: ('a286', None, ('item286',))

```
(over a couple of hours).

Adding a delay in flush, or timing it so that it will run flush in the middle (or any time after the beginning) of the rollup interval would make this harder to hit, but I don't see a way to make it actually correct without using synchronization primitives (e.g. locks).

**Comment by edkorthof** [comment]
This is actually a python script to illustrate -- I've reproduced with a couple of different versions; I'm not sure if the targeted sleep for the threads doing add helps -- I think the main thing is having enough load on the host system that the python process has to share. When it does so, it can be interrupted in between the lines of python code above, causing this error.


[example.txt](https://github.com/DataDog/datadogpy/files/2994601/example.txt)

**Comment by lyoungblood** [comment]
This makes ThreadStats unusable in Python Lambda functions. After a few hours of container re-use eventually none of your Lambdas will execute due to thread exhaustion.

**Comment by zippolyte** [comment]
Should be fixed by #370, closing.
